feature/gribi/ate_tests/ordering_ack_test/ordering_ack_test.go

```diff
@@ -0,0 +1,503 @@
+func awaitTimeout(ctx context.Context, c *fluent.GRIBIClient, t testing.TB) error {
+	subctx, cancel := context.WithTimeout(ctx, awaitDuration)
+	defer cancel()
+	return c.Await(subctx, t)
+}
+
+// testArgs holds the objects needed by a test case.
+type testArgs struct {
+	ctx context.Context
+	c   *fluent.GRIBIClient
+	dut *ondatra.DUTDevice
+	ate *ondatra.ATEDevice
+	top *ondatra.ATETopology
+}
+
+// testCaseFunc describes a test case function.
+type testCaseFunc func(t *testing.T, args *testArgs)
+
+// testModifyNHG configures a NextHopGroup referencing a NextHop.
+func testModifyNHG(t *testing.T, args *testArgs) {
+	args.c.Modify().AddEntry(t,
+		fluent.NextHopEntry().
+			WithNetworkInstance(instance).
+			WithIndex(nhIndex).
+			WithIPAddress(ateDst.IPv4),
+		fluent.NextHopGroupEntry().
+			WithNetworkInstance(instance).
+			WithID(nhgIndex).
+			AddNextHop(nhIndex, nhWeight),
+	)
+	if err := awaitTimeout(args.ctx, args.c, t); err != nil {
+		t.Errorf("Await got error for ModifyRequest: %v", err)
+	}
+
+	res := args.c.Results(t)
+	chk.HasResult(t, res,
+		fluent.OperationResult().
+			WithOperationID(1).
+			WithOperationType(constants.Add).
+			WithNextHopOperation(nhIndex).
+			WithProgrammingResult(wantInstalled).
+			AsResult(),
+	)
+	chk.HasResult(t, res,
+		fluent.OperationResult().
+			WithOperationID(2).
+			WithOperationType(constants.Add).
+			WithNextHopGroupOperation(nhgIndex).
+			WithProgrammingResult(wantInstalled).
+			AsResult(),
+	)
+
+	t.Run("Telemetry", func(t *testing.T) {
+		if !*checkTelemetry {
+			t.Skip()
+		}
+		nhgNhPath := args.dut.Telemetry().NetworkInstance(instance).Afts().NextHopGroup(nhgIndex).NextHop(nhIndex)
+		if got, want := nhgNhPath.Index().Get(t), uint64(nhIndex); got != want {
+			t.Errorf("next-hop-group/next-hop/state/index got %d, want %d", got, want)
+		}
+		if got, want := nhgNhPath.Weight().Get(t), uint64(nhWeight); got != want {
+			t.Errorf("next-hop-group/next-hop/state/weight got %d, want %d", got, want)
+		}
+	})
+}
+
+// testModifyIPv4NHG configures a ModifyRequest with a NextHop and an IPv4Entry before a
+// NextHopGroup which is invalid due to the forward reference.
+func testModifyIPv4NHG(t *testing.T, args *testArgs) {
+	args.c.Modify().AddEntry(t,
+		fluent.NextHopEntry().
+			WithNetworkInstance(instance).
+			WithIndex(nhIndex).
+			WithIPAddress(ateDst.IPv4),
+		fluent.IPv4Entry().
+			WithNetworkInstance(instance).
+			WithPrefix(ateDstNetCIDR).
+			WithNextHopGroup(nhgIndex),
+		fluent.NextHopGroupEntry().
+			WithNetworkInstance(instance).
+			WithID(nhgIndex).
+			AddNextHop(nhIndex, nhWeight),
+	)
+	if err := awaitTimeout(args.ctx, args.c, t); err != nil {
+		t.Fatalf("Await got error for ModifyRequest: %v", err)
+	}
+
+	res := args.c.Results(t)
+	chk.HasResult(t, res,
+		fluent.OperationResult().
+			WithOperationID(2).
+			WithOperationType(constants.Add).
+			WithIPv4Operation(ateDstNetCIDR).
+			WithProgrammingResult(fluent.ProgrammingFailed).
```

Comment:
We are seeing different expectation regarding out of order when those are sent in one modify request vs multiple modify requests. gribigo [compliance testcase](https://github.com/openconfig/gribigo/blob/a471e9b7c03e2d121a3b5d23e44c17fb37514c47/compliance/compliance.go#L542) is expecting forward reference to work when multiple modify requests are sent. should the behavior be consistent? @xw-g

Reply:
sure. thanks